tests/test_cache/test_remove.py

```diff
@@ -0,0 +1,86 @@
+import json
+import os
+import time
+import pytest
+
+from threading import Thread
+
+from fades import cache
+
+
+def test_missing_file(tmp_file):
+    venvscache = cache.VEnvsCache(tmp_file)
+    venvscache.remove('missing/path')
+
+    lines = venvscache._read_cache()
+    assert lines == []
+
+
+def test_missing_env_in_cache(tmp_file):
+    venvscache = cache.VEnvsCache(tmp_file)
+    options = {'foo': 'bar'}
+    venvscache.store('installed', {'env_path': 'some/path'}, 'interpreter', options=options)
+    lines = venvscache._read_cache()
+    assert len(lines) == 1
+
+    venvscache.remove('some/path')
+
+    lines = venvscache._read_cache()
+    assert lines == []
+
+
+def test_preserve_cache_data_ordering(tmp_file):
+    venvscache = cache.VEnvsCache(tmp_file)
+    # store 3 venvs
+    options = {'foo': 'bar'}
+    venvscache.store('installed1', {'env_path': 'path/env1'}, 'interpreter', options=options)
+    venvscache.store('installed2', {'env_path': 'path/env2'}, 'interpreter', options=options)
+    venvscache.store('installed3', {'env_path': 'path/env3'}, 'interpreter', options=options)
+
+    venvscache.remove('path/env2')
+
+    lines = venvscache._read_cache()
+    assert len(lines) == 2
+    assert json.loads(lines[0]).get('metadata').get('env_path') == 'path/env1'
+    assert json.loads(lines[1]).get('metadata').get('env_path') == 'path/env3'
+
+
+@pytest.mark.skip(reason="I dont know why is not working with pytest")
```

Comment:
```bash
tmp_file = '/private/var/folders/lm/2g_qb49141x31rvmqs4pf3fw0000gn/T/pytest-of-eduardoenriquez/pytest-19/test0/testfile', mocker = <pytest_mock.MockFixture object at 0x11024a6a0>

    def test_lock_cache_for_remove(tmp_file, mocker):
        venvscache = cache.VEnvsCache(tmp_file)
        # store 3 venvs
        options = {'foo': 'bar'}
        venvscache.store('installed1', {'env_path': 'path/env1'}, 'interpreter', options=options)
        venvscache.store('installed2', {'env_path': 'path/env2'}, 'interpreter', options=options)
        venvscache.store('installed3', {'env_path': 'path/env3'}, 'interpreter', options=options)
    
        # patch _write_cache so it emulates a slow write during which
        # another process managed to modify the cache file before the
        # first process finished writing the modified cache data
        original_write_cache = venvscache._write_cache
        p = mocker.patch('fades.cache.VEnvsCache._write_cache')
        mock_write_cache = p.start()
    
        t1 = Thread(target=venvscache.remove, args=('path/env1',))
    
        def slow_write_cache(*args, **kwargs):
            p.stop()
            t1.start()
            # wait to ensure t1 thread must wait for lock to be released
            time.sleep(0.01)
            original_write_cache(*args, **kwargs)
    
        mock_write_cache.side_effect = slow_write_cache
    
        # just a sanity check
        assert not os.path.exists(venvscache.filepath + '.lock')
        # remove a virtualenv from the cache
        venvscache.remove('path/env2')
>       t1.join()

tests/test_cache/test_remove.py:78: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <Thread(Thread-7, initial)>, timeout = None

    def join(self, timeout=None):
        """Wait until the thread terminates.
    
        This blocks the calling thread until the thread whose join() method is
        called terminates -- either normally or through an unhandled exception
        or until the optional timeout occurs.
    
        When the timeout argument is present and not None, it should be a
        floating point number specifying a timeout for the operation in seconds
        (or fractions thereof). As join() always returns None, you must call
        is_alive() after join() to decide whether a timeout happened -- if the
        thread is still alive, the join() call timed out.
    
        When the timeout argument is not present or None, the operation will
        block until the thread terminates.
    
        A thread can be join()ed many times.
    
        join() raises a RuntimeError if an attempt is made to join the current
        thread as that would cause a deadlock. It is also an error to join() a
        thread before it has been started and attempts to do so raises the same
        exception.
    
        """
        if not self._initialized:
            raise RuntimeError("Thread.__init__() not called")
        if not self._started.is_set():
>           raise RuntimeError("cannot join thread before it is started")
E           RuntimeError: cannot join thread before it is started

/usr/local/Cellar/python/3.7.3/Frameworks/Python.framework/Versions/3.7/lib/python3.7/threading.py:1027: RuntimeError
```

Reply:
Fixed in last commit